builds/fabric8-che/plugins/keycloak-plugin-server/src/main/java/com/redhat/che/keycloak/server/KeycloakAuthenticationFilter.java

```diff
@@ -83,13 +87,55 @@ private boolean isSystemStateRequest(String requestURI) {
         return requestURI.endsWith("/api/system/state");
     }
 
-    private boolean isInternalRequest(String authHeader) {
-        return "Internal".equals(authHeader);
+    /**
+     * @param requestURI
+     * @return true if request is made against endpoint which provides keycloak
+     *         status (enabled / disabled), false otherwise
+     */
+    private boolean isKeycloakSettingsRequest(String requestURI) {
+        return requestURI.endsWith("/keycloak/settings");
+    }
+
+    /**
+     * @param authHeader
+     * @return true if "Authorization" header has valid openshift token that is
+     *         used for communication between wsagent and wsmaster, false
+     *         otherwise
+     */
+    private boolean isWorkspaceAgentRequest(String authHeader) {
+        if (authHeader != null && authHeader.startsWith("Wsagent")) {
+            String token = authHeader.replaceFirst("Wsagent ", "");
+            return isValidToken(token);
+        }
+        return false;
     }
 
     private boolean isWebsocketRequest(String requestURI, String requestScheme) {
         return requestURI.endsWith("/ws") || requestURI.endsWith("/eventbus") || requestScheme.equals("ws")
                 || requestScheme.equals("wss") || requestURI.contains("/websocket/")
                 || requestURI.endsWith("/token/user");
     }
+
+    /**
+     * OpenShift default service account token is used in "Authorization" header
+     * for communication between wsagent and wsmaster. The method checks if the
+     * token is valid by fetching namespace info. Examples of valid che
+     * namespaces are eclipse-che, username-che, ibuziuk-che etc.
+     * 
+     * @param token
+     * @return true if openshift token is valid, false otherwise
+     */
+    private boolean isValidToken(final String token) {
+        LOG.debug("Validating workspace agent token");
+        Config config = new ConfigBuilder().withOauthToken(token).build();
+        try (OpenShiftClient client = new DefaultOpenShiftClient(config)) {
+            String namespace = client.getConfiguration().getNamespace();
+            LOG.debug("Validating the token against namespace '{}'", namespace);
+            return namespace.contains("-che");
```

Comment:
Will this have the issue that I can use my token, prepended with Wsagent, to access someone else's workspace? 

e.g. If I get my OSO token, afaict the code above will return true if I pass the auth header `Wsagent <my token>`, since my namespace will be something like `<username>-che`?